(maint) add a batch processing script that only requires the SHA/tag …

speedofdark opened the pull request:
…for puppet-agent.

ekinanp commented on matchbatch.sh:
```diff
@@ -0,0 +1,95 @@
+# need the puppet agent revision as seed
+git checkout ${1}
+
+componentRevMap=$(getComponentRevMap)
```
repoRevMap might be a better name? b/c the puppet-agent repo isn't a component

speedofdark replied:
Updated...

ekinanp commented on matchbatch.sh:
```diff
@@ -0,0 +1,95 @@
+componentRevMap="puppet-agent:${1} ${componentRevMap}"
+
+for currentPair in ${componentRevMap}; do
+	component=$(echo -n ${currentPair} | cut -d: -f1)
```
repo here instead of component?

speedofdark replied:
Updated...

ekinanp reviewed (commented):
Some really small comments, not super important.

Might not be a bad idea to have this take some environment variables? E.g. WORKSPACE for the workspace containing the agent repo and the components (default to PWD if it isn't provided). 

I think you can also do the cloning part here by having componentRevMap include the component URL and clone from that if the component does not exist in the workspace. We can maybe have another environment variable that dictates whether we use the private agent or the public agent that we can use to calculate the agent repo's URL for cloning (e.g. USE_PRIVATE_AGENT) [can default to using the public puppet-agent).

Although environment variables are hacky, these would be an easy way to include some of my feature requests without having to do a whole lot of big modifications at once. What do you think?

ekinanp commented:
Oh, it might be good to indicate that the script's for the puppet-agent and its components in the filename? E.g. maybe `pa-matchbatch.sh`?

caseywilliams commented on ticketmatch.rb:
```diff
@@ -284,15 +292,15 @@ def inspect
 if jira_issues['issues'].nil?
   say("JIRA returned no results for project '#{jira_project_name}' and fix version '#{jira_project_fixed_version}'")
-  exit(status=1)
+  exit 0
```
I think the matchbatch additions could be split into their own commit so that they're distinct from the fixes to existing code (which look good), but I don't think it's that important. As a note for later, it would be nice to make this script check for all variables at once when run in non-interactive mode so that it can report all missing variables before `abort`ing, but again, not a big deal right now.

speedofdark replied:
I agree with checking things en masse; short to medium term change.

speedofdark commented:
I like a number of these ideas. Would the simplest way to get the first set
of changes (the script and repo name change ones) into this same PR be a
commit --amend?

On Tue, Apr 17, 2018 at 10:45 AM, ekinanp <notifications@github.com> wrote:

> Oh, it might be good to indicate that the script's for the puppet-agent
> and its components in the filename? E.g. maybe pa-matchbatch.sh?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/puppetlabs/ticketmatch/pull/6#issuecomment-382081399>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgHR1_5Io-np0gqZRPI7kUryb8i_YHQ_ks5tpipEgaJpZM4TYut9>
> .
>

caseywilliams commented on matchbatch.sh:
```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env bash
```
I'd kind of prefer to see all ruby in this repo, since we're already using it for ticketmatch proper, but I am glad that this agent-specific stuff is separated from the main script, and this looks useful. Whatever works.

speedofdark replied:
Thats the medium term plan.

ekinanp commented:
@speedofdark Yup, `git commit --amend` is correct.

speedofdark committed:
(maint) add a batch processing script that only requires the SHA/tag for puppet_agent.

- can use the environment var WORKSPACE to point to a puppet-agent repo; defaults to PWD.
- if a component repo is not present, it will be cloned from the component.json URL.

- cleaned up component parsing
- handle gem-based variations (resource_api for now)
- added all current components to be processable
- lots of BASH-idomatic stuff
- cleaned up output processing
- not finding things in the specified Git range is no longer terminal
- please review the README for additional information

- version matching now takes merge-ups into account; README has details
- no longer requires empty directories, will fetch instead of clone as needed
- provide means of overriding JIRA 'fixed in' versions along with version summary

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,93 @@
+[ -z "${puppetAgentBaseRev}" ] && { echo "Error: must specify the puppet-agent revision to start from."; exit 1; }
+
+cd ${WORKSPACE:-${PWD}}
+[ $(basename ${PWD}) == 'puppet-agent' ] || { echo "Error: must be in a puppet-agent directory"; exit 1; }
```
I think you should also clone the agent if it does not exist. You can use git@github.com:puppetlabs/puppet-agent.git as the URL. Can add this to the repoRevMap.

speedofdark replied:
I like the idea of getting puppet-agent if you don't have it already, but
you need it to generate the initial repoRevMap, so it can't be inside the
loop. Chicken and egg problem. Switching to the desired tag can be, but not
the initial clone.

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,93 @@
+
+# get rev hashes in the form [component]|[to_rev]|[url]
+getComponentRevMap() {
+	for componentName in $(for componentFile in $(grep -lv refs/tags configs/components/*.json); do grep -l puppetlabs ${componentFile}; done); do 
```
A comment here clarifying the grep stuff would be nice. I think what's going on here is you're looking for components pinned to tags, and of those, filtering out the ones owned by puppetlabs right?

speedofdark replied:
Exactly.

speedofdark commented:
Exactly.

On Tue, Apr 17, 2018 at 3:53 PM, ekinanp <notifications@github.com> wrote:

> *@ekinanp* commented on this pull request.
> ------------------------------
>
> In pa_matchbatch.sh
> <https://github.com/puppetlabs/ticketmatch/pull/6#discussion_r182265799>:
>
> > @@ -0,0 +1,93 @@
> +#!/usr/bin/env bash
> +set -e
> +
> +# get rev hashes in the form [component]|[to_rev]|[url]
> +getComponentRevMap() {
> +	for componentName in $(for componentFile in $(grep -lv refs/tags configs/components/*.json); do grep -l puppetlabs ${componentFile}; done); do
>
> A comment here clarifying the grep stuff would be nice. I think what's
> going on here is you're looking for components pinned to tags, and of
> those, filtering out the ones owned by puppetlabs right?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/puppetlabs/ticketmatch/pull/6#pullrequestreview-113029105>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgHR17Vg2tEi2i3lNK6Ck3wv9mXSRh3yks5tpnJggaJpZM4TYut9>
> .
>

speedofdark commented:
I like the idea of getting puppet-agent if you don't have it already, but
you need it to generate the initial repoRevMap, so it can't be inside the
loop. Chicken and egg problem. Switching to the desired tag can be, but not
the initial clone.

On Tue, Apr 17, 2018 at 3:48 PM, ekinanp <notifications@github.com> wrote:

> *@ekinanp* commented on this pull request.
> ------------------------------
>
> In pa_matchbatch.sh
> <https://github.com/puppetlabs/ticketmatch/pull/6#discussion_r182264924>:
>
> > +		pxp-agent) echo pxp-agent
> +		;;
> +		puppet-agent) echo puppet-agent
> +		;;
> +		*) (>&2 echo "Error: need to add JIRA fixed-in version mapping for '${1}'.")
> +			exit 1
> +		;;
> +	esac
> +}
> +
> +# main
> +puppetAgentBaseRev=${1}
> +[ -z "${puppetAgentBaseRev}" ] && { echo "Error: must specify the puppet-agent revision to start from."; exit 1; }
> +
> +cd ${WORKSPACE:-${PWD}}
> +[ $(basename ${PWD}) == 'puppet-agent' ] || { echo "Error: must be in a puppet-agent directory"; exit 1; }
>
> I think you should also clone the agent if it does not exist. You can use
> git@github.com:puppetlabs/puppet-agent.git as the URL. Can add this to
> the repoRevMap.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/puppetlabs/ticketmatch/pull/6#pullrequestreview-113028155>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgHR1_8VYN_kki8y_DQqFUFB30ze6JIdks5tpnEygaJpZM4TYut9>
> .
>

ekinanp commented:
Oh right, that makes sense. Then I think it should be OK to clone it before init'ing the repoRevMap?

speedofdark commented:
Yup. Otherwise there's nothing to init it with :)...

caseywilliams commented:
Hooray for the readme updates - still LGTM :+1:

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,149 @@
+baseDir=${WORKSPACE:-${PWD}}
+[[ -d ${baseDir} ]] || { echo "Error: '${baseDir}' -- no such directory."; exit 1; }
+cd ${baseDir}
+[[ $(ls -A | wc -l) -eq 0 ]] || { echo "Error: '${PWD}' is not empty."; exit 1; }
```
I would remove this check, because that makes it so people can't run and then re-run the script several times. Instead, might be better to clone the puppet-agent from the start if it does not exist and go from there

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,149 @@
+			git log -1 --no-merges --oneline --grep='Release prep' | sed -e "s/.*prep for v\(.*$\)/\1/"
+		;;
+		*)
+			git log -1 --no-merges --oneline -E --grep='\(packaging\) .* version' | sed -e "s/.*'\(.*\)'.*/\1/"
```
So there is a really stupid subtlety here and that's merge-ups. With merge-ups, we can have a version bump from a previous maint branch "override" (in git log) the version bump in our current branch/ref. I found this out after a test run of your script, where I passed it the 5.5.x agent, which uses the 5.5.x puppet as its component, but the reported version was a 4.10.x version (b/c the version bump for 4.10.x got merged up and overwrote the one for 5.5.x in git log [it is reported before it])

It might be better to just read the version from the repo directly. There's some code you can copy-paste here for that: https://github.com/puppetlabs/ci-job-configs/blob/master/resources/scripts/vanagon-helpers.sh#L372-L448

^ The "{{" brace stuff is b/c of interpolation; all of the "{{" or "}}" you see should be replaced with "{" or "}" respectively if you want to use it in your script. (You can probably ignore the other cases in the read_repo_version function b/c those were added by different teams).

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,149 @@
+
+# get desired revision of puppet-agent in place
+echo "Note: cloning puppet-agent..."
+git clone --quiet ${PUPPET_AGENT_URL}
```
If you are going to be using the read_repo_version thing I linked in an earlier comment, then you might want to specify the private forks as part of the special cases (or maybe parse out the "-private" part altogether as a first-pass filter in the routine).

ekinanp reviewed (changes requested):
Some comments, but generally looks good. I did a test-run of the code here, and aside from that read_version subtlety I mentioned, things seem to work!

I like the PUPPET_AGENT_URL environment variable.

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,186 @@
+ORIGIN_PATH=${ORIGIN_PATH:-origin}
+
+# where version overrides are specified
+overridesFile="/tmp/version_overrides.txt"
```
Might be good to have an ENV variable for this w/ a default value?

ekinanp commented on pa_matchbatch.sh:
```diff
@@ -0,0 +1,186 @@
+TICKETMATCH_PATH=${TICKETMATCH_PATH:-$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )}
+
+# where origin actually is
+ORIGIN_PATH=${ORIGIN_PATH:-origin}
```
I would rename this to `FETCH_REMOTE`, `ORIGIN_PATH` is a bit weird b/c the clone URL will almost always be the `origin` remote.

speedofdark commented:
Done and done.

On Mon, Apr 30, 2018 at 2:47 PM, ekinanp <notifications@github.com> wrote:

> *@ekinanp* commented on this pull request.
> ------------------------------
>
> In pa_matchbatch.sh
> <https://github.com/puppetlabs/ticketmatch/pull/6#discussion_r185119989>:
>
> > @@ -0,0 +1,186 @@
> +#!/usr/bin/env bash
> +set -e
> +unset CDPATH
> +
> +# where to find the agent
> +puppet_agent_repo="git@github.com:puppetlabs/puppet-agent.git"
> +PUPPET_AGENT_URL=${PUPPET_AGENT_URL:-${puppet_agent_repo}}
> +
> +# where to find the ticketmatch.rb script
> +TICKETMATCH_PATH=${TICKETMATCH_PATH:-$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )}
> +
> +# where origin actually is
> +ORIGIN_PATH=${ORIGIN_PATH:-origin}
>
> I would rename this to FETCH_REMOTE, ORIGIN_PATH is a bit weird b/c the
> clone URL will almost always be the origin remote.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/puppetlabs/ticketmatch/pull/6#pullrequestreview-116448827>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AgHR1zyiL9eD1g_isnMnWRA4cxty-287ks5tt4ZngaJpZM4TYut9>
> .
>
